### ISSUE TITLE
cache() could pass through items while buffering for replay

### ISSUE THREAD

**Issue by davidmoten** [ReactiveX/RxJava#3026]
Using 1.0.12 this bit of code indicates that `cache` reads its whole source before emitting anything:

```java
Observable.just(1, 2)
    .delay(1, TimeUnit.SECONDS, Schedulers.immediate())
    .cache()
    .repeat(2)
    .doOnNext(System.out::println)
    .subscribe()
```

Running this we get
```
<wait two seconds>
1
2
1
2
```
I'd like to get:
```
<wait one second>
1
<wait one second>
2
1
2
```
@akarnokd you've done some work on `cache` lately that I haven't looked at closely. Could an improvement like this be incorporated?

**Comment by akarnokd** [comment]
Whit the current head and this code:
```java
Observable.just(1, 2)
        .delay(1, TimeUnit.SECONDS, Schedulers.immediate())
        //.take(2)
        .cache()
        .repeat(2)
        .timeInterval()
        .doOnNext(System.out::println)
        .subscribe();
```

I get
```
TimeInterval [intervalInMilliseconds=1016, value=1]
TimeInterval [intervalInMilliseconds=1001, value=2]
TimeInterval [intervalInMilliseconds=1015, value=1]
TimeInterval [intervalInMilliseconds=0, value=2]
```

The reason for the delay before the second `1` is that delay() delays the `onCompleted()` event by a second as well so the repeat can't resubscribe that early. If I uncomment the `take(2)` before the cache, that's going to trigger an earlier completion.

**Comment by davidmoten** [comment]
Thanks @akarnokd. Pleased to hear it.

**Comment by akarnokd** [comment]
This is now officially fixed in 1.0.14.
